### DISCUSSION TITLE
Using System.Numerics

### DISCUSSION THREAD

**Discussion by thetoeb** [opentk/opentk#1688] [Ideas]
I noticed that the build-in vector/matrix types don't support SIMD. Also some minor features are missing, like matrix/quaternion lerp for example. That is all included in the System.Numerics namesspace, and the byte layout of theses structs are exactly the same as the build-in ones. So why we don't just use the System.Numerics? I use it in my current project with OpenTk and its works nicely, but i always have to use "Unsafe.As" the whole time. What are (besides the work that it takes) the downsides of this approach?

**Comment by NogginBops** [comment]
OpenTK math predates `System.Numerics` so that is one reason why it exists. Another reason is that we implement many more math types than `System.Numerics`.

`System.Numerics` also uses SIMD to accelerate singular math operations, this is generally not the strength of SIMD (for some simple operations such a vector addition it works fine, but for more complicated tasks it doesn't work as well). SIMD is good for accelerating multiple operations in parallel.
This doesn't really matter if you aren't doing the SIMD implementations yourself, so using `System.Numerics` can be a nice "default" perf boost.

OpenTK 4 doesn't bind OpenGL functions using System.Numerics as the bindings generator hard to work with. OpenTK 5 will use both OpenTK math and System.Numerics.

If you could open an issue detailing some of the missing math functions I'd be happy to implement them.
